MiniDB/DataBase.cs

```diff
@@ -0,0 +1,1170 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Threading;
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace MiniDB
+{
+    /// <summary>
+    /// Database Template: template class for persisten observable collection with undo/redo
+    /// </summary>
+    /// <typeparam name="T">The class type to create an observable collection of (must by a child class of DatabaseObject)</typeparam>
+    public class DataBase<T> : ObservableCollection<T>, IDisposable where T : DatabaseObject
```

Comment:
It's late, so I don't have time to properly review this behemoth. So here's the highlights:

- It looks like your Encryption subclass could be redesigned as a database save strategy, which could go a good distance in simplifying logic in this class (no worrying about subclasses) and that class (no worrying about base classes).

- _Any_ time you need a destructor, add a comment to your code explaining what horrible things can go wrong if the destructor isn't called, and add a Debug.Assert to the destructor to check that it got called in the way you expected.